### PR TITLE
Align KTO with DPO: Remove null_ref_context

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Remove `null_ref_context`.

This PR refactors how the code handles the case when there is no explicit reference model (`ref_model is None`), especially for PEFT (Parameter-Efficient Fine-Tuning) models. The main change is the removal of the `null_ref_context` context manager in favor of directly inlining its logic where needed. This makes the code more straightforward and localized, reducing indirection.

Part of:
- #4786

### Changes

**Refactoring context management for PEFT models in KTO:**

* Removed the `null_ref_context` context manager and replaced its usage by inlining the adapter-switching logic directly where reference log probabilities are computed.

* Updated the `compute_ref_log_probs` and `_compute_loss` methods to directly check for PEFT models and handle adapter switching inline, improving clarity and reducing unnecessary abstraction.

These changes simplify the codebase and make the handling of reference log probabilities for PEFT models more explicit and easier to follow.

---

> [!NOTE]
> **Medium Risk**
> Changes how KTO selects the reference policy during training and precompute paths; behavior should match prior logic for PEFT/no-ref_model setups, but mistakes here would skew KTO rewards/loss.
> 
> **Overview**
> Removes the **`null_ref_context`** helper from **`KTOTrainer`** and inlines how reference log-probs are computed when there is no separate **`ref_model`**, matching the DPO trainer pattern (#4786).
> 
> In **`compute_ref_log_probs`**, a **`ref_model is None`** path now branches explicitly: PEFT models run completion (and optional KL) forwards under **`use_adapter`** with the **`ref`** adapter when present; non-PEFT models call **`self.model`** directly without the old context wrapper.
> 
> In **`_compute_loss`**, live reference log-probs use the same inlined **`use_adapter`** block only when **`is_peft_model(self.model)`** and **`ref_model is None`**; otherwise they still come from **`self.ref_model`**. Precomputed **`ref_logps`** behavior is unchanged.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 50009cb33af6fddb0dcbc0cba49fc5ec4faa4616. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -1077,8 +1079,9 @@ def _compute_loss(self, model, inputs, return_outputs):
                 ref_KL_logps = None
         else:
             with torch.no_grad():
-                if self.ref_model is None:
-                    with self.null_ref_context():
+                if is_peft_model(self.model) and self.ref_model is None:
+                    model = self.accelerator.unwrap_model(self.model)
+                    with use_adapter(model, adapter_name="ref" if "ref" in model.peft_config else None):
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:1084]: I think this claim is impossible:
- self.ref_model is None  only when:
  - is_peft_model(self.model) is True, OR
  - args.precompute_ref_log_probs is True
- the mentioned code is the `else` branch of `if self.precompute_ref_logps`
- therefore, self.precompute_ref_logps is False, and then
  - either is_peft_model(self.model) is True AND self.ref_model is None
  - or is_peft_model(self.model) is False AND self.ref_model is NOT None

[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/kto/kto_trainer.py:1084]: > when the model is not a PEFT model and no reference model is provided

yes, when the model is not a PEFT model and no reference model is provided, it means that you've pre-computed the ref logps

**Review by qgallouedec — APPROVED** [review]
